Extension header: amend const-correctness of `p_args` parameters

Bromeon opened the pull request:
Addendum to https://github.com/godotengine/godot/pull/67751.

Since we likely don't intend to modify the elements (pointers) of the arguments array, I'd suggest to const-qualify _both_ value and pointer. This way, the API signals that no mutable access to the elements is required.

The `p_args` type would then evolve as follows:

| | Parameter type | Resolved type |
|---|---|---|
| Original | `const GDNativeTypePtr*` | `void* const*` |
| After #67751 | `GDNativeConstTypePtr*` | `void const**` |
| After this PR | `const GDExtensionConstTypePtr*` | `void const* const*` |

(analogous for variant pointers).

Now, there's a small problem that the entire engine uses `const void**` and `const Variant**` all over the place, and it would be a huge refactoring the change all occurrences, and possibly hurt readability (at least without `typedef`). This is why changes here are limited to `gdextension_interface.h/cpp` files.

In other words: it would likely be correct to use double-const all over the engine code, but this wasn't done before #67751, either. The `(const Variant**)` C-casts contains a hidden `const_cast` in that sense -- this might be something to fix for later.

touilleMan reviewed (approved):
lgtm

akien-mga commented:
godot-cpp would need to have that same change merged first to be able to pass the CI checks here.

akien-mga commented:
> godot-cpp would need to have that same change merged first to be able to pass the CI checks here.

Merged the godot-cpp counterpart, a rebase should fix this one.

Bromeon committed:
Extension header: amend const correctness of `p_args` parameters

A while ago, argument arrays were passed as const GDNativeTypePtr* (void* const*)
This was changed to GDNativeConstTypePtr* (void const**), adding const on the value but removing it on the pointer level.
This commit changes argument types to const GDExtensionConstTypePtr* (void const* const*).

Besides object pointers, the same change is applied to variant pointers.

Bromeon commented:
Rebased.

akien-mga commented:
Thanks!
